lib/src/Dscp/DB/SQLite/Queries.hs

```diff
@@ -155,6 +224,184 @@ getStudentTransactions student = do
         where      student_addr = ?
     |]
 
+-- | Returns list of transaction blocks along with block-proof of a student since given moment.
+getProvenStudentTransactionsSince :: DBM m => Id Student -> UTCTime -> m [(MerkleProof PrivateTx, [(Word32, PrivateTx)])]
+getProvenStudentTransactionsSince studentId sinceTime = do
+    transaction $ do
+        -- Contains `(tx, idx, blockId)` map.
+        txsBlockList <- getTxsBlockMap
+
+        -- Bake `blockId -> [(tx, idx)]` map.
+        let txsBlockMap = groupToAssocWith (_tibBlockId, _tibTx) txsBlockList
+
+        results <- forM txsBlockMap $ \(blockId, transactions) -> do
+            blockData <- getBlockData blockId
+
+            let tree    = _bdTree blockData
+                indiced = [(idx, tx) | TxWithIdx tx idx <- transactions]
+                mapping = Map.fromList indiced
+                pruned  = fillEmptyMerkleTree mapping tree
+
+            return (pruned, indiced)
+
+        return [(proof, txs) | (Just proof, txs) <- results]
+
+  where
+    groupToAssocWith :: Ord k => Ord k => (a -> k, a -> v) -> [a] -> [(k, [v])]
+    groupToAssocWith (key, value) =
+        Map.toList . foldr' push Map.empty
+      where
+        push a = Map.insertWith (++) (key a) [value a]
```

Comment:
Why not `Map.insertWith (flip (:)) (key a) (value a)` ?

Reply:
It will reverse order (I don't know if we need to preserve it).

Reply:
I used `flip (++)`, because `Map.insertWith` expects `v -> v -> v` there.